core/src/main/java/com/datastax/oss/driver/internal/core/session/DefaultSession.java

```diff
@@ -337,39 +345,57 @@ private void processDistanceEvent(DistanceEvent event) {
 
     private void processStateEvent(NodeStateEvent event) {
       assert adminExecutor.inEventLoop();
-      // no need to check closeWasCalled, because we stop listening for events one closed
+      // no need to check closeWasCalled, because we stop listening for events once closed
       DefaultNode node = event.node;
+      NodeState oldState = event.oldState;
       NodeState newState = event.newState;
       if (pending.containsKey(node)) {
         pendingStateEvents.put(node, event);
       } else if (newState == NodeState.FORCED_DOWN) {
         ChannelPool pool = pools.remove(node);
         if (pool != null) {
-          LOG.debug("[{}] {} became FORCED_DOWN, destroying pool", logPrefix, node);
+          LOG.debug("[{}] {} was FORCED_DOWN, destroying pool", logPrefix, node);
           pool.closeAsync()
               .exceptionally(
                   error -> {
                     LOG.warn("[{}] Error closing pool", logPrefix, error);
                     return null;
                   });
         }
-      } else if (newState == NodeState.UP) {
-        ChannelPool pool = pools.get(node);
-        if (pool == null) {
-          LOG.debug("[{}] {} came back UP and no pool found, initializing it", logPrefix, node);
-          CompletionStage<ChannelPool> poolFuture =
-              channelPoolFactory.init(node, keyspace, node.getDistance(), context, logPrefix);
-          pending.put(node, poolFuture);
-          poolFuture
-              .thenAcceptAsync(this::onPoolInitialized, adminExecutor)
-              .exceptionally(UncaughtExceptions::log);
-        } else {
-          LOG.debug("[{}] {} came back UP, triggering pool reconnection", logPrefix, node);
-          pool.reconnectNow();
+      } else if (oldState == NodeState.FORCED_DOWN
+          && newState == NodeState.UP
+          && node.getDistance() != NodeDistance.IGNORED) {
+        LOG.debug("[{}] {} was forced back UP, initializing pool", logPrefix, node);
+        createOrReconnectPool(node);
+      }
+    }
+
+    private void onTopologyEvent(TopologyEvent event) {
+      assert adminExecutor.inEventLoop();
+      if (event.type == TopologyEvent.Type.SUGGEST_UP) {
+        Node node = context.metadataManager().getMetadata().getNodes().get(event.address);
+        if (node.getDistance() != NodeDistance.IGNORED) {
+          LOG.debug(
+              "[{}] Received a SUGGEST_UP event for {}, reconnecting pool now", logPrefix, node);
+          createOrReconnectPool(node);
```

Comment:
Actually if a node is up it should never have a null pool, at worst an empty and reconnecting pool. But I figured it wouldn't hurt to recreate the pool if it's null.
Same thing in reverse in `processStateEvent` (if the node is forced back up, it should always create, never reconnect).